### PR TITLE
Add MemberDescriptor tests and fix bugs

### LINKED ISSUE #38984: MemberDescriptor.Attributes throws NRE if null attribute in attributes
E.g.
```cs
new SubMemberDescriptor("Name", new Attribute[] { null }).Attributes
```

### LINKED ISSUE #38985: new MemberDescriptor((MemberDescriptor)null) throws NRE
Should be ANE

### LINKED ISSUE #38986: new MemberDescriptor(new MemberDescriptorWithNullName()) throws NRE
Should "just work"

### LINKED ISSUE #38987: MemberDescriptor.FillAttributes(null) throws NRE
Should throw ANE or be a nop

### PR DESCRIPTION
**Pull request by hughbe**
Fixes #38987
Fixes #38986
Fixes #38985
Fixes #38984

### CODE REVIEW

**Comment by hughbe** [comment]
azp run corefx-ci

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs [hunk 1]
```diff
@@ -327,9 +345,8 @@ private void FilterAttributesIfNeeded()
                     {
                         FillAttributes(list);
                     }
-                    catch (Exception e)
+                    catch (Exception)
                     {
-                        Debug.Fail($"{_name}>>{e}");
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs]: Why are you removing this?

Comment (hughbe) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs]: Can be triggered and the tests I wrote did

Comment (safern) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs]: How would it be triggered? What is the exception thrown?

File: src/System.ComponentModel.TypeConverter/tests/MemberDescriptorTests.cs [hunk 2]
```diff
@@ -2,105 +2,1547 @@
...
+            yield return new object[] { "name", attributes4, new Attribute[] { attributes4[0], attributes4[2] } };
+        }
+
+        [Theory]
+        [MemberData(nameof(Ctor_String_Attributes_TestData))]
+        public void Ctor_String_Attributes(string name, Attribute[] attributes, Attribute[] expected)
+        {
+            var descriptor = new SubMemberDescriptor(name, attributes);
+            Assert.Equal(expected, descriptor.Attributes.Cast<Attribute>());
+            Assert.Same(descriptor.Attributes, descriptor.Attributes);
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MemberDescriptorTests.cs]: Was your intent to compare the descriptor property against itself?

Comment (safern) [reply src/System.ComponentModel.TypeConverter/tests/MemberDescriptorTests.cs]: I guess because Same does a ref comparison?

Comment (maryamariyan) [reply src/System.ComponentModel.TypeConverter/tests/MemberDescriptorTests.cs]: nit: there is a lot less occurrences of Assert.Same as opposed to Assert.Equal in the repo (approx 600 vs 4000).
I believe both Same and Equal will always result in the same outcome, so might want to switch to Assert.Equal instead.

cc: @danmosemsft

Comment (danmosemsft) [reply src/System.ComponentModel.TypeConverter/tests/MemberDescriptorTests.cs]: I would use Equal unless you specifically want reference equality. That is unlikely especially for eg strings

Comment (hughbe) [reply src/System.ComponentModel.TypeConverter/tests/MemberDescriptorTests.cs]: Use of Same is deliberate across this PR. Making sure the we have reference equality. This is especially important here as if `Attributes` returned a new instance each time, things would be very very bad

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs (hunk 1 above)
Comment (hughbe) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs]: This is in the test `Attributes_GetWithThrowingFillAttributes_ReturnsExpected`

```cs
[Fact]
public void Attributes_GetWithThrowingFillAttributes_ReturnsExpected()
{
    var descriptor = new ThrowingFillAttributesMemberDescriptor("Name");
    Assert.Equal(new Attribute[] { descriptor.Attribute }, descriptor.Attributes.Cast<Attribute>());
}

[Fact]
public void AttributeArray_GetWithThrowingFillAttributes_ReturnsExpected()
{
    var descriptor = new ThrowingFillAttributesMemberDescriptor("Name");
    Assert.Equal(new Attribute[] { descriptor.Attribute }, descriptor.AttributeArray);
}

private class ThrowingFillAttributesMemberDescriptor : MemberDescriptor
{
    public ThrowingFillAttributesMemberDescriptor(string name) : base(name)
    {
    }

    public Attribute Attribute { get; } = new MockAttribute1();

    public new Attribute[] AttributeArray
    {
        get => base.AttributeArray;
        set => base.AttributeArray = value;
    }

    protected override void FillAttributes(IList attributeList)
    {
        attributeList.Add(Attribute);
        throw new NotImplementedException();
    }
}
```

Comment (hughbe) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs]: Old behaviour: release no exception thrown - it is caught. debug assert
New behaviour: release/debug no exception thrown - it is caught

File: src/System.ComponentModel.TypeConverter/tests/MemberDescriptorTests.cs (hunk 2 above)
Comment (danmosemsft) [reply src/System.ComponentModel.TypeConverter/tests/MemberDescriptorTests.cs]: Fair enough

**Pull request merged**